### PR TITLE
Added in-memory MetadataService. Closes #784

### LINKED ISSUE #784: Implement Metadata Service API with in-memory persistence

### PR DESCRIPTION
**Pull request by cjllanwarne**
An in-memory implementation of the MetadataService

### CODE REVIEW

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala [hunk 1]
```diff
@@ -63,15 +63,43 @@ object MetadataServiceActor {
...
+      )
+      case None => List.empty
+    }
+
+    val possibleFilters: List[Option[MetadataEvent => Boolean]] = List(
+      metadataQuery.workflowId map { queryId => (event: MetadataEvent) => event.key.workflowId == queryId },
+      metadataQuery.key map { queryKey => (event: MetadataEvent) => event.key.key == queryKey }
+    ) ++ possibleJobQueryFilters
+
+    val filtersToUse: List[MetadataEvent => Boolean] = possibleFilters collect { case Some(f) => f }
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: `possibleFilters.flatten`

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala [hunk 2]
```diff
@@ -63,15 +63,43 @@ object MetadataServiceActor {
...
       sender ! result
   }
+
+  private def filterByQuery(metadataQuery: MetadataQuery): Seq[MetadataEvent] = {
+
+    val possibleJobQueryFilters: List[Option[MetadataEvent => Boolean]] = metadataQuery.jobKey match {
+      case Some(jobKeyQuery) => List(
+        jobKeyQuery.callFqn map { queryCallFqn => (event: MetadataEvent) => event.key.jobKey.isDefined && event.key.jobKey.get.callFqn == queryCallFqn },
+        jobKeyQuery.index map { queryIndex => (event: MetadataEvent) => event.key.jobKey.isDefined && event.key.jobKey.get.index == queryIndex },
+        jobKeyQuery.attempt map { queryAttempt => (event: MetadataEvent) => event.key.jobKey.isDefined && event.key.jobKey.get.attempt == queryAttempt }
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: the inner things here would probably be better as pattern matches

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala [hunk 3]
```diff
@@ -32,7 +32,7 @@ object MetadataServiceActor {
   object MetadataQuery {
     def forWorkflow(workflowId: WorkflowId) = MetadataQuery(Option(workflowId), None, None)
     def forJob(workflowId: WorkflowId, jobKey: MetadataJobKey) = MetadataQuery(Option(workflowId), Option(MetadataQueryJobKey.forMetadataJobKey(jobKey)), None)
-    def forKey(key: String) = MetadataQuery(None, None, Option(key))
+    def forKey(key: MetadataKey) = MetadataQuery(Option(key.workflowId), key.jobKey map { MetadataQueryJobKey.forMetadataJobKey }, Option(key.key))
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: those curlies are probably superfluous

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala [hunk 4]
```diff
@@ -63,15 +63,43 @@ object MetadataServiceActor {
...
+
+  var journal = List.empty[MetadataEvent]
+
   def receive = {
-    case PutMetadataAction(event) =>
-      val result: MetadataPutAcknowledgement = ???
+    case action@PutMetadataAction(event) =>
+      journal = journal :+ event
+      val result: MetadataPutAcknowledgement = MetadataPutAcknowledgement(action)
       sender ! result
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: inline `result`

File: engine/src/test/scala/cromwell/services/MetadataServiceSpec.scala [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+import akka.pattern.ask
+import com.typesafe.config.ConfigFactory
+import cromwell.core.WorkflowId
+import cromwell.services.MetadataServiceActor._
+
+class MetadataServiceSpec extends CromwellServicesSpec {
+
+  val config = ConfigFactory.load("{}")
+  val Actor = actorSystem.actorOf(MetadataServiceActor.props(config, config))
```
Comment (mcovarr) [inline engine/src/test/scala/cromwell/services/MetadataServiceSpec.scala]: lowercase

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala [hunk 6]
```diff
@@ -63,15 +63,43 @@ object MetadataServiceActor {
...
+
+    val possibleFilters: List[Option[MetadataEvent => Boolean]] = List(
+      metadataQuery.workflowId map { queryId => (event: MetadataEvent) => event.key.workflowId == queryId },
+      metadataQuery.key map { queryKey => (event: MetadataEvent) => event.key.key == queryKey }
+    ) ++ possibleJobQueryFilters
+
+    val filtersToUse: List[MetadataEvent => Boolean] = possibleFilters collect { case Some(f) => f }
+
+    // Make sure that every filter which is being requested applies to this event:
+    if (filtersToUse.isEmpty) journal else journal filter { event => filtersToUse forall { _(event) } }
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: not sure you need this `if`

**Comment by mcovarr** [comment]
:+1:

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala [hunk 7]
```diff
@@ -63,15 +63,43 @@ object MetadataServiceActor {
 }
 
 case class MetadataServiceActor(serviceConfig: Config, globalConfig: Config) extends Actor {
+
+  var journal = List.empty[MetadataEvent]
```
Comment (gauravs90) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: Could be private I suppose.

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala (hunk 2 above)
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: Which do you mean? The `map` or the `isDefined && ...` ?

Comment (mcovarr) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: not a huge improvement but this seems a bit better:

```scala
        jobKeyQuery.attempt map { queryAttempt => (event: MetadataEvent) => event.key.jobKey.exists(_.attempt == queryAttempt) }
```

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala [hunk 8]
```diff
@@ -63,15 +63,40 @@ object MetadataServiceActor {
...
     case GetAllMetadataAction(workflowId) =>
-      val result: MetadataLookupResponse = ???
-      sender ! result
-    case GetMetadataQueryAction(MetadataQuery(workflowId, jobKey, key)) =>
-      val result: MetadataLookupResponse = ???
-      sender ! result
+      val query = MetadataQuery(Option(workflowId), None, None)
+      val filtered = filterByQuery(query)
+      sender ! MetadataLookupResponse(query, filtered)
+    case GetMetadataQueryAction(query@MetadataQuery(workflowId, jobKey, key)) =>
```
Comment (gauravs90) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: Not a strong opinion here, but `workflowId`, `jobKey` and `key` are all unused here and probably good candidates for a wildcard `_` to emphasize it's non-usefulness.

**Comment by gauravs90** [comment]
👍

**Pull request merged**